### PR TITLE
Add `mlflow.diffusers` flavor for diffusion model LoRA adapters

### LINKED ISSUE #22122: [FR] Native flavor for diffusion model adapters (LoRA, LoKr, LoHa)
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

MLflow has no way to log and register diffusion model adapters (LoRA, LoKr, LoHa) as first-class models. These are the standard way to fine-tune Stable Diffusion, SDXL, and Flux models, but they don't fit any existing MLflow flavor because they're not standalone models — they're lightweight weight adapters that get applied on top of a base model at inference time.

A native `diffusers-adapter` flavor (or extending the existing `transformers` flavor) would let users log, version, and deploy LoRA models the same way they do with any other ML model in MLflow.

### Motivation

> #### What is the use case for this feature?

Fine-tuning diffusion models with LoRA is one of the most common workflows in generative AI right now. Tools like ai-toolkit, kohya-ss, and diffusers all produce LoRA weights as `.safetensors` files. Users need to track which LoRA version came from which training run, compare them, and eventually serve them.

> #### Why is this use case valuable to support for MLflow users in general?

LoRA fine-tuning is how most teams customize image generation models. The ecosystem is huge — Civitai alone hosts hundreds of thousands of LoRA models. Anyone using MLflow for diffusion model training hits the same wall: you can track metrics and log images, but the actual model output (the LoRA weights) is just a plain artifact with no model registry support.

> #### Why is this use case valuable to support for your project(s) or organization?

We're building an MLflow integration for ostris/ai-toolkit (LoRA training for Stable Diffusion, SDXL, Flux). Everything works except model registration — the LoRA weights don't show up in the Models section because there's no flavor that understands adapter weights.

> #### Why is it currently difficult to achieve this use case?

You can't even register a LoRA as a model without a workaround. `mlflow.register_model()` on a raw artifact fails on MLflow 3.x with `Unable to find a logged_model` — it requires a `log_model()` call, not just `log_artifact()`. So you're forced to create a pyfunc stub just to get into the Model Registry:

```python
class LoRAStub(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        raise NotImplementedError("Use diffusers to load this LoRA adapter")

mlflow.pyfunc.log_model(
    name="lora_model",
    python_model=LoRAStub(),
    artifacts={"lora_weights": "output/lora.safetensors"},
    registered_model_name="my-lora",
)
```

This works for registration and lineage but the model can't actually be served. Beyond that, even if you tried to write a real `predict()`:

- You have to write a custom model class that handles loading the base model + applying the adapter
- The base model (often 5-10GB) can't reasonably be bundled as an artifact, so you end up storing a reference string and hoping it resolves at load time
- Different adapter formats (safetensors, kohya-ss, diffusers) need different loading logic
- The `predict()` signature doesn't map well to image generation (prompts in, images out, with dozens of generation params like guidance_scale, num_inference_steps, etc.)
- No standardized way to declare which base model the adapter was trained on

A native flavor would handle all of this with a clean API like:
```python
mlflow.diffusers.log_adapter(
    adapter_path="output/lora.safetensors",
    base_model="black-forest-labs/FLUX.1-dev",
    adapter_type="lora",
    artifact_path="lora_model",
)
```

### Details

A native flavor would need to handle:

- **Adapter types**: LoRA (most common), LoKr, LoHa, full fine-tune diffs
- **Formats**: safetensors (standard), kohya-ss state dict format, diffusers format
- **Base model reference**: store which base model the adapter was trained on without bundling the full model weights
- **Loading**: use `diffusers` library to load base model + apply adapter via `pipe.load_lora_weights()`
- **Inference signature**: prompts + generation params → images
- **Model card metadata**: adapter rank, training dataset, base model, etc.

This could either be a new `mlflow.diffusers_adapter` flavor or an extension of the existing `mlflow.transformers` flavor since diffusers is part of the Hugging Face ecosystem.

### What machine learning domain(s) is this feature request about?

- `domain/deep-learning`

### What area(s) of MLflow is this feature request about?

- `area/models`

### PR DESCRIPTION
**Pull request by Rasaboun**
### Related Issues/PRs

Closes #22122

### What changes are proposed in this pull request?

Add a new `mlflow.diffusers` flavor for logging, versioning, and serving diffusion model LoRA adapters.

- `save_model` / `log_model` / `load_model` following the standard MLflow flavor pattern
- `load_model()` returns a `DiffusersAdapterModel` with `load_pipeline()` for direct inference
- pyfunc `predict()` support: prompt string in, PNG bytes out
- Default signature: `prompt (string) → image (binary)` with params (`num_inference_steps`, `guidance_scale`, `height`, `width`)
- Base model stored as a reference string in flavor config, adapter weights copied as artifacts
- Single-file adapters normalized to `pytorch_lora_weights.safetensors` for compatibility with any training framework
- Model-agnostic: uses `DiffusionPipeline.from_pretrained()` which auto-resolves the pipeline class

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

32 tests covering save/load/log roundtrip, signature, metadata, filename normalization, pip requirements, error handling, and pyfunc predict() with input validation.

Manually tested end-to-end with a real FLUX.2 Klein 4B LoRA fine-tune on GPU including checkpoint logging and artifact verification through the MLflow API and UI.

### Does this PR require documentation update?

- [ ] No.
- [x] Yes. I've updated:
  - [x] Examples
  - [ ] API references
  - [x] Instructions

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.
- [ ] Yes. Please link the corresponding PR or explain how you plan to update it.

#### Is this a user-facing change?

- [ ] No.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add `mlflow.diffusers` flavor for logging and serving diffusion model LoRA adapters. Enables `save_model`, `log_model`, `load_model`, and pyfunc inference for adapters trained with diffusers/PEFT.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by Rasaboun** [comment]
Quick walkthrough of the 5 commits:

**1. refactor** — Pulled `get_latest_commit_for_repo` and `is_valid_hf_repo_id` out of `mlflow/transformers/hub_utils.py` into a shared `mlflow/utils/huggingface_utils.py` since the new diffusers flavor needs them too. The old module re-exports for backward compat.

**2. feat** — The flavor itself. Follows the standard MLflow pattern. Two areas that need careful eyes:

At save time we pin the base model's HF Hub commit SHA for reproducibility. If the hub is unreachable (offline, air-gapped), pinning is skipped gracefully instead of failing the save.

PEFT saves adapters as `adapter_model.safetensors` but diffusers' `load_lora_weights()` expects `pytorch_lora_weights.safetensors`. For single files we rename to the standard name. For directories with companion files (like `adapter_config.json`) we preserve the layout and record the actual filename as `weight_name` in the flavor config — otherwise models break in offline mode where diffusers can't query the Hub to guess the filename.

The pyfunc wrapper lazily loads the pipeline on first `predict()` with double-checked locking for concurrent serving.

**3. test** — Covers save/load roundtrips, weight_name edge cases (multi-file, PEFT adapters), hidden file filtering, all predict input types, error paths, revision pinning, etc.

**4. docs** — Docs page with prerequisites (`diffusers >= 0.37.0`), getting started examples, signature table, and a demo script that works without GPU.

**5. fix(transformers)** — One-liner unrelated to diffusers. `persist_pretrained_model` was missing a `pop(MODEL_LOCAL_BASE)` next to the existing `pop(MODEL_REVISION)`, so models saved with `base_model_path` would ignore persisted weights at load time.

### E2E GPU testing

Tested end-to-end on Modal with real GPU inference:

- **FLUX.2 Klein 4B** — fake LoRA adapter → `log_model` → `load_model` → `load_pipeline` → generate 512x512 image → pyfunc `predict()` returns valid PNG bytes. All 5 steps pass.

- **Real ai-toolkit fine-tune** — trained a FLUX.2 Klein 4B LoRA for 5 steps on a synthetic dataset using [ai-toolkit](https://github.com/ostris/ai-toolkit) with MLflow logging enabled. The training run logs 41 params (lr, optimizer, lora_rank, etc.), per-step loss metrics, training sample images, and the training dataset. After training, the LoRA adapter is logged with `mlflow.diffusers.log_model()` into the same run. Full roundtrip verified: `load_model()` → `load_pipeline()` → generate image.

Both tests ran on A100 GPUs via Modal. The MLflow UI (3.11.1.dev0) correctly displays all metrics, parameters, sample images, datasets, and model artifacts including the diffusers flavor config with base model revision pinning.

**Comment by Rasaboun** [comment]
@TomeHirata The PR is ready for review and fully tested e2e and I have add unit test

**Comment by Rasaboun** [comment]
@TomeHirata Up

File: mlflow/diffusers/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        elif isinstance(data, dict):
+            if "prompt" not in data:
+                raise MlflowException(
+                    f"Input dict must contain a 'prompt' key. Got keys: {list(data.keys())}"
+                )
+            prompts = data["prompt"]
+            if isinstance(prompts, str):
+                prompts = [prompts]
+            elif isinstance(prompts, list):
+                prompts = self._flatten_prompts(prompts)
```
Comment (TomeHirata) [inline mlflow/diffusers/wrapper.py:107]: If `data["prompt"]` is neither a `str` nor a `list` (e.g. a numpy scalar, integer, or `pd.Series`), neither branch matches and `prompts` is passed as-is to the pipeline, which will produce a confusing error deep inside diffusers. Consider adding an explicit type-check and raising a `MlflowException` with a descriptive message:

```suggestion
            prompts = data["prompt"]
            if isinstance(prompts, str):
                prompts = [prompts]
            elif isinstance(prompts, list):
                prompts = self._flatten_prompts(prompts)
            else:
                raise MlflowException(
                    f"'prompt' value must be a string or list of strings, got {type(prompts).__name__}."
                )
```

File: mlflow/diffusers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+def log_model(
+    adapter_path,
+    base_model_id,
+    adapter_type: Literal["lora"] = "lora",
+    artifact_path: str | None = None,
+    conda_env=None,
+    code_paths=None,
+    registered_model_name=None,
+    signature: ModelSignature = None,
+    input_example: ModelInputExample = None,
```
Comment (TomeHirata) [inline mlflow/diffusers/__init__.py]: `signature` and `input_example` are typed without `| None`, inconsistent with `save_model` which has the correct `ModelSignature | None = None` annotation. 

```suggestion
    signature: ModelSignature | None = None,
    input_example: ModelInputExample | None = None,
```

File: mlflow/transformers/hub_utils.py [hunk 3]
```diff
@@ -1,82 +1,4 @@
-import functools
-import logging
-import os
-import time
+# Re-export from the shared location for backward compatibility.
+from mlflow.utils.huggingface_utils import get_latest_commit_for_repo, is_valid_hf_repo_id
```
Comment (TomeHirata) [inline mlflow/transformers/hub_utils.py]: it's fine not to keep the backward compatibility here since these are not public APIs

File: mlflow/diffusers/__init__.py [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+    except Exception as e:
+        raise MlflowException.invalid_parameter_value(
+            f"File is not a valid safetensors file: {file_path}. Error: {e}"
+        ) from e
+
+
+def _detect_device(device=None):
+    import torch
+
+    from mlflow.environment_variables import MLFLOW_DEFAULT_PREDICTION_DEVICE
```
Comment (TomeHirata) [inline mlflow/diffusers/__init__.py]: Can we move this to the module level?

File: mlflow/diffusers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+
+    Args:
+        adapter_path: Path to the adapter weights. Can be a single .safetensors file
+            or a directory containing adapter files. Single files and directories
+            containing a single safetensors file are normalized to
+            ``pytorch_lora_weights.safetensors`` to match the convention expected
+            by ``load_lora_weights()``. Directories with multiple weight files
+            are copied as-is.
+        path: Local path where the model is to be saved.
+        base_model_id: HuggingFace model ID or local path of the base diffusion model
```
Comment (TomeHirata) [inline mlflow/diffusers/__init__.py]: Shall we use `base_model` since it accepts a local path?

File: mlflow/diffusers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+            A ``DiffusionPipeline`` with LoRA weights applied.
+        """
+        from diffusers import DiffusionPipeline
+
+        device = _detect_device(kwargs.pop("device", None))
+        kwargs.setdefault("torch_dtype", "auto")
+        if self.base_model_revision and "revision" not in kwargs:
+            kwargs["revision"] = self.base_model_revision
+
+        pipe = DiffusionPipeline.from_pretrained(self.base_model_id, **kwargs)
```
Comment (TomeHirata) [inline mlflow/diffusers/__init__.py]: How do we support the case where the local model path is different between loading time and logging time?

File: mlflow/diffusers/__init__.py (hunk 4 above)
Comment (Rasaboun) [reply mlflow/diffusers/__init__.py]: Yes

File: mlflow/diffusers/__init__.py (hunk 5 above)
Comment (Rasaboun) [reply mlflow/diffusers/__init__.py]: Sure, base_model is more accurate I'll rename it

File: mlflow/diffusers/__init__.py (hunk 6 above)
Comment (Rasaboun) [reply mlflow/diffusers/__init__.py]: Today if the local path doesn't exist at load time, from_pretrained crashes with a confusing diffusers error. I've added :
- load_pipeline(base_model=...) override so users can pass a different path or Hub ID
- model_config={"base_model": "..."} support for pyfunc/serving
- Both paths wrap the OSError with a clear message pointing to the override

For portable models, users should use HuggingFace Hub IDs. Local paths work for single-machine workflows, and the override covers the case where the path changes

**Comment by Rasaboun** [comment]
@TomeHirata Thanks for the review ! I've addressed all comments let me know if the approach works for you, especially the base_model rename and the local path handling

File: mlflow/diffusers/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        weight_name = self._flavor_conf.get("weight_name")
+        lora_kwargs = {}
+        if weight_name:
+            lora_kwargs["weight_name"] = weight_name
+
+        _logger.info("Loading base pipeline: %s", base_model)
+        try:
+            pipe = DiffusionPipeline.from_pretrained(base_model, **load_kwargs)
+        except OSError as e:
+            raise MlflowException(
```
Comment (TomeHirata) [inline mlflow/diffusers/wrapper.py:49]: can we add error_code=INVALID_PARAMETER_VALUE?

**Review by TomeHirata — APPROVED** [review]
LGTM, can you fix lint?

**Comment by Rasaboun** [comment]
@TomeHirata Done

> LGTM, can you fix lint?

**Comment by Rasaboun** [comment]
@TomeHirata I have rebase from the main branch so it will fix the github workflow errors
